docs/programming-oak.md

```diff
@@ -90,39 +58,147 @@ macro converts the provided function body into an
 
-For Nodes that act as gRPC servers (the normal "front door" for an Oak
-Application), the easiest way to use a gRPC service implementation is to:
+For an Oak Application that exposes itself as a gRPC server (the normal "front
+door" for an Oak Application), the easiest way to set things up is to define two
+nodes:
 
-- call the
-  [`oak::grpc::server::init()`](https://project-oak.github.io/oak/sdk/doc/oak/grpc/server/fn.init.html)
-  helper from the Oak SDK to set up a
-  [gRPC server pseudo-Node](concepts.md#pseudo-nodes) and acquire a channel to
-  receive gRPC invocations on
-- wrap the Node `struct` in the automatically generated `Dispatcher`, as
-  described in the [next section](#generated-grpc-service-code)
-- pass the `Dispatcher` and the channel handle to the
-  [`run_event_loop()`](https://project-oak.github.io/oak/sdk/doc/oak/fn.run_event_loop.html)
-  function.
+- a main Node
+- a gRPC service handler Node
+
+We will describe them in the next sections, starting with the gRPC service
+handler Node.
+
+#### gRPC Service Handler Node
+
+This Node implements the logic related to handling
+[`GrpcInvocation`](https://project-oak.github.io/oak/sdk/doc/oak/proto/oak/invocation/struct.GrpcInvocation.html)
+messages coming from a gRPC server pseudo-Node.
+
+In order to implement it:
+
+- define a `struct` holding the internal state of the gRPC service, or just an
+  empty `struct` if no state is needed.
+- implement the auto-generated gRPC service trait for the Node handler `struct`.
+- define an
+  [`impl_dispatcher!`](https://project-oak.github.io/oak/sdk/doc/oak/macro.impl_dispatcher.html)
+  entry, specifying the appropriate auto-generated gRPC dispatcher, in order to
+  automatically make the handler `struct` implement the
+  [`CommandHandler`](https://project-oak.github.io/oak/sdk/doc/oak/trait.CommandHandler.html)
+  trait.
 
 <!-- prettier-ignore-start -->
-[embedmd]:# (../examples/translator/module/rust/src/lib.rs Rust /oak::entrypoint!\(grpc_oak_main/ /^}/)
+[embedmd]:# (../examples/translator/module/rust/src/lib.rs Rust /oak::entrypoint_command_handler!\(handler/ /^}/)
 ```Rust
-oak::entrypoint!(grpc_oak_main<ConfigMap> => |_receiver| {
-    oak::logger::init_default();
-    let grpc_channel =
-        oak::grpc::server::init("[::]:8080").expect("could not create gRPC server pseudo-Node");
-    oak::run_command_loop(Node, grpc_channel.iter());
+oak::entrypoint_command_handler!(handler => Handler);
+oak::impl_dispatcher!(impl Handler : TranslatorDispatcher);
+
+#[derive(Default)]
+struct Handler;
+
+impl Translator for Handler {
+    fn translate(&mut self, req: TranslateRequest) -> grpc::Result<TranslateResponse> {
+        info!(
+            "attempt to translate '{}' from {} to {}",
+            req.text, req.from_lang, req.to_lang
+        );
+        let mut rsp = TranslateResponse::default();
+        rsp.translated_text = match req.from_lang.as_str() {
```

Comment:
There's a lot of code here that doesn't show anything relevant; is it possible to find a smaller example or to split into a start + end section with "// snipped" in between?  

(Fine as-is if it's not easy)

Reply:
Yeah I thought about that, I was actually going to start a slack thread about that, it feels like some of this documentation would fit better as inline comments on the actual code, and we just point to the source file from the readme. Let me start a thread there and let's see where it goes.